libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,387 @@
+    let derPrivKey = key.seckey.rsakey.getBytes.get
+    let pemPrivKey: string = pemEncode(derPrivKey, "PRIVATE KEY")
+    token.sign(pemPrivKey)
+    $token.toFlattenedJson()
+
+proc requestRegister*(
+    self: ACMEApi, key: KeyPair
+): Future[ACMERegisterResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  let registerRequest = ACMERegisterRequest(termsOfServiceAgreed: true)
+  handleError("acmeRegister"):
+    let payload = await self.createSignedAcmeRequest(
+      self.directory.newAccount, registerRequest, key, needsJwk = true
+    )
+    let rawResponse = await HttpClientRequestRef
+    .post(
+      self.session, self.directory.newAccount, body = payload, headers = ACMEHttpHeaders
+    )
+    .get()
+    .send()
+    let body = await rawResponse.getResponseBody()
+    let headers = rawResponse.headers
+    ACMERegisterResponse(
+      status: body["status"].getStr, kid: headers.getString("location")
+    )
+
+proc requestNewOrder(
+    self: ACMEApi, domains: seq[string], key: KeyPair, kid: Kid
+): Future[(string, ACMEChallengeResponse)] {.
+    async: (raises: [ACMEError, CancelledError])
+.} =
+  # request challenge from ACME server
+  let orderRequest = ACMEChallengeRequest(
+    identifiers: domains.mapIt(ACMEChallengeIdentifier(`type`: "dns", value: it))
+  )
+  handleError("requestNewOrder"):
+    let payload = await self.createSignedAcmeRequest(
+      self.directory.newOrder, orderRequest, key, kid = Opt.some(kid)
+    )
+    let rawResponse = await HttpClientRequestRef
+    .post(
+      self.session, self.directory.newOrder, body = payload, headers = ACMEHttpHeaders
+    )
+    .get()
+    .send()
+    let body = await rawResponse.getResponseBody()
+    let headers = rawResponse.headers
+    (headers.getString("location"), body.to(ACMEChallengeResponse))
+
+proc requestAuthorizations(
+    self: ACMEApi, authorizations: seq[string], key: KeyPair, kid: Kid
+): Future[ACMEAuthorizationsResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  handleError("requestAuthorizations"):
+    let rawResponse =
+      await HttpClientRequestRef.get(self.session, authorizations[0]).get().send()
+    let body = await rawResponse.getResponseBody()
+    body.to(ACMEAuthorizationsResponse)
+
+proc requestChallenge*(
+    self: ACMEApi, domains: seq[string], key: KeyPair, kid: Kid
+): Future[ACMEChallengeResponseWrapper] {.async: (raises: [ACMEError, CancelledError]).} =
+  let (orderURL, challengeResponse) = await self.requestNewOrder(domains, key, kid)
+
+  if challengeResponse.authorizations.len() == 0:
+    raise newException(ACMEError, "Authorizations field is empty")
+
+  let authorizationsResponse =
+    await self.requestAuthorizations(challengeResponse.authorizations, key, kid)
+
+  let dns01 = authorizationsResponse.challenges.filterIt(it.`type` == "dns-01")[0]
+
+  return ACMEChallengeResponseWrapper(
+    finalizeURL: challengeResponse.finalize, orderURL: orderURL, dns01: dns01
+  )
```

Comment:
```suggestion
  let authorizationsResponse =
    await self.requestAuthorizations(challengeResponse.authorizations, key, kid)

  return ACMEChallengeResponseWrapper(
    finalizeURL: challengeResponse.finalize, 
    orderURL: orderURL, 
    dns01: authorizationsResponse.challenges.filterIt(it.`type` == "dns-01")[0]
  )
```

^ might be opinionated suggestion but 
if `var dns01` can be initialized in one line
then better directly initialize struct field